### PR TITLE
Adds the initial support for AOTInductor model and interface

### PR DESCRIPTION
**Pull request by chenyang78**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103922

This PR combines the C++ code for the AOTInductor's model and interface with
Bin Bao's changes to AOTInductor codegen.

It adds a number of AOTInductor C interfaces that can be used by an
inference runtime. Under the hood of the interfaces, the model code generated by
the AOTInductor's codegen is wrapped into a class, AOTInductorModel, which
manages tensors and run the model inference.

On top of AOTInductorModel, we provide one more abstract layer,
AOTInductorModelContainer, which allows the user to have multiple inference
runs concurrently for the same model.

This PR also adjusts the compilation options for AOT codegen, particularly
some fbcode-related changes such as libs to be linked and header-file search
paths.

Note that this is the very first version of the AOTInductor model and interface,
so many features (e.g. dynamic shape) are incomplete. We will support those
missing features in in future PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225

### CODE REVIEW

File: torch/_inductor/aot_cpp_include/aot_inductor_interface.cpp
Comment (jansel) [inline torch/_inductor/aot_cpp_include/aot_inductor_interface.cpp:1]: For this (and the other cpp files) to exist when people `pip install torch` I believe we need to update setup.py:
https://github.com/pytorch/pytorch/blob/0330f67b222bb58dcc3b47941338086e7538e704/setup.py#L1252

File: test/cpp/aot_inductor/test.py [hunk 1]
```diff
@@ -17,6 +17,9 @@ def forward(self, x, y):
 x = torch.randn((32, 64), device="cuda")
 y = torch.randn((32, 64), device="cuda")
 
+# FIXME: re-enable dynamic shape after we add dynamic shape support to the
+# AOTInductor runtime
+torch._dynamo.config.dynamic_shapes = False
```
[the remark this answers is not in the document]
Comment (chenyang78) [reply test/cpp/aot_inductor/test.py]: Fixed. Thanks.

File: torch/_inductor/aot_cpp_include/aot_inductor_interface.cpp
Comment (chenyang78) [reply torch/_inductor/aot_cpp_include/aot_inductor_interface.cpp:1]: Yes, good point. Fixed.

File: test/cpp/aot_inductor/CMakeLists.txt [hunk 2]
```diff
@@ -30,6 +30,10 @@ target_link_libraries(test_aot_inductor PRIVATE
   ${CMAKE_CURRENT_BINARY_DIR}/libaot_inductor_output.so
 )
 
+target_include_directories(test_aot_inductor PRIVATE
```
Comment (desertfire) [inline test/cpp/aot_inductor/CMakeLists.txt:33]: With your change to setup.py, I suspect we don't need this now?

File: torch/_inductor/codecache.py [hunk 3]
```diff
@@ -584,6 +587,11 @@ def cpp_flags():
 
 def optimization_flags():
     base_flags = "-O3 -ffast-math -fno-finite-math-only"
+    if config.is_fbcode():
```
Comment (desertfire) [inline torch/_inductor/codecache.py:590]: You need to coordinate with Brian on this.

File: torch/_inductor/codecache.py [hunk 4]
```diff
@@ -614,22 +629,32 @@ def get_include_and_linking_paths(
         # to do to enable OMP build on darwin where PyTorch is built with IOMP
         # and we need a way to link to what PyTorch links.
         ipaths = cpp_extension.include_paths(cuda) + [sysconfig.get_path("include")]
+        if aot_mode:
+            ipaths += [os.path.join(_TORCH_PATH, "_inductor", "aot_cpp_include")]
```
Comment (desertfire) [inline torch/_inductor/codecache.py:633]: This probably also needs to be updated after the setup.py change.

File: torch/_inductor/codecache.py [hunk 5]
```diff
@@ -614,22 +629,32 @@ def get_include_and_linking_paths(
...
+            else:
+                libs += ["c10_cuda", "cuda", "torch_cuda"]
     else:
         # Note - this is effectively a header only inclusion. Usage of some header files may result in
         # symbol not found, if those header files require a library.
         # For those cases, include the lpath and libs command as we do for pytorch above.
         # This approach allows us to only pay for what we use.
         ipaths = cpp_extension.include_paths(cuda) + [sysconfig.get_path("include")]
+        if aot_mode:
+            ipaths += [os.path.join(_INDUCTOR_PATH, "aot_cpp_include")]
```
Comment (desertfire) [inline torch/_inductor/codecache.py:657]: This else branch is not tested, that's why _INDUCTOR_PATH is not defined but not caught. Honestly we don't care about this else branch, so probably ok to drop the change here.

File: torch/_inductor/codegen/wrapper.py [hunk 6]
```diff
@@ -1086,19 +1163,31 @@ def codegen_device(self, device):
...
+        for idx, output in enumerate(V.graph.graph_outputs):
+            if isinstance(output, (ir.NoneAsConstantBuffer, ir.ShapeAsConstantBuffer)):
+                continue
+            if buffer == output.data:
+                output_idx = idx
+                break
+        if output_idx is not None and V.graph.aot_mode:
+            # In aot_mode, output buffers are managed by the AOT runtime.
+            return (
+                f"at::Tensor {buffer.get_name()} = outputs[{output_idx}]{self.ending}"
```
Comment (desertfire) [inline torch/_inductor/codegen/wrapper.py:1176]: How do we guarantee the passed-in output buffers are valid? So basically what is our protocol with the runtime?

File: setup.py [hunk 7]
```diff
@@ -1194,6 +1194,8 @@ def main():
         'include/THH/generic/*.h',
         'include/sleef.h',
         "_inductor/codegen/*.h",
+        "_inductor/aot_cpp_include/*.cpp",
+        "_inductor/aot_cpp_include/*.h",
```
[the remark this answers is not in the document]
Comment (chenyang78) [reply setup.py:1198]: Done. Thanks.

File: test/cpp/aot_inductor/CMakeLists.txt (hunk 2 above)
Comment (chenyang78) [reply test/cpp/aot_inductor/CMakeLists.txt:33]: We still need this. ```torch/_inductor``` is not in the header-file search paths.

File: torch/_inductor/codecache.py (hunk 3 above)
Comment (chenyang78) [reply torch/_inductor/codecache.py:597]: Yes, I will check with Brian. Thanks.

File: torch/_inductor/codecache.py (hunk 4 above)
Comment (chenyang78) [reply torch/_inductor/codecache.py]: Yes, changed.

File: torch/_inductor/codecache.py (hunk 5 above)
Comment (chenyang78) [reply torch/_inductor/codecache.py]: Done. Thanks.

File: torch/_inductor/codegen/wrapper.py (hunk 6 above)
Comment (chenyang78) [reply torch/_inductor/codegen/wrapper.py:1176]: The runtime will first invoke ```AOTInductorModelContainerGetMaxInputShape``` to query the output tensor shapes before allocating the right amount of memory.

**Review by desertfire — APPROVED** [review]

**Comment by jansel** [comment]
The merge button doesn't work for ghstack.

**Pull request merged**